### PR TITLE
non-global DSL feature

### PR DESCRIPTION
**Pull request by tarvit**
I believe it would be nice to have a possibility to use Kemal::DSL methods in non-global namespace

### CODE REVIEW

**Comment by tarvit** [comment]
Here is a concept example
```ruby

# load Kemal without DSL in a global context
require "kemal/main"

class MyApp < Kemal::Application
  # some very app specific code if needed
end

class OkController < Kemal::Controller
  def process
    # access to HTTP::Server::Context
    puts context.inspect
    "OK!"
  end
end

# You can call `MyApp#configure_routes` multiple times,
# so different route scopes can be defined in different files 
MyApp.configure_routes do |routes|
  routes.add "get", "/ok", OkController
end

MyApp.run
```

With Sockets
```ruby
require "kemal/main"

class MyApp < Kemal::Application
end

# A simple wrapper for a socket api, 
# with ability to incapsulate logic in a single class
class HelloController < Kemal::SocketController
  def process
    socket.send "Hello from Kemal!"

    socket.on_message do |message|
      socket.send "Echo back from server #{message}"
    end

    socket.on_close do
      puts "Closing socket"
    end
  end
end

MyApp.configure_routes do |routes|
  routes.socket "/ws", HelloController
end

MyApp.run
```

**Comment by sdogruyol** [comment]
This is an interesting approach, I'd like to take some time to throughly review 👍

**Comment by tarvit** [comment]
All changes are compatible with a current Kemal DSL

**Comment by sdogruyol** [comment]
Does this support having multiple Kemal applications in same process?

File: src/kemal/app/controller.cr [hunk 1]
```diff
@@ -0,0 +1,10 @@
+module Kemal
+  abstract class Controller
```
Comment (sdogruyol) [inline src/kemal/app/controller.cr]: I'd like to call this `Action` instead of `Controller` to avoid confusion.

File: src/kemal/app/socket_controller.cr [hunk 2]
```diff
@@ -0,0 +1,10 @@
+module Kemal
+  abstract class SocketController
```
Comment (sdogruyol) [inline src/kemal/app/socket_controller.cr]: Same here `SocketAction`

**Comment by tarvit** [comment]
> Does this support having multiple Kemal applications in same process?
Yes. But you should call `Kemal.run` instead of calling `MyFirstApp.run` && `MySecondApp.run`
I am not sure about how useful this approach could be. Because currently `Kemal` has some central singleton instances and other global stuff. So I we need some really separate multiple instances I think we need to redesign few objects in Kemal core.

File: src/kemal/app/controller.cr (hunk 1 above)
Comment (tarvit) [inline src/kemal/app/controller.cr]: I think it is a good idea. I will rename that

**Comment by sdogruyol** [comment]
Thanks for your effort @tarvit, I think this is a good step forward.

File: spec/app/application_layer_spec.cr [hunk 3]
```diff
@@ -0,0 +1,28 @@
+require "../spec_helper"
+
+class MyApp < Kemal::Application
+end
+
+class OkController < Kemal::Action
```
Comment (sdogruyol) [inline spec/app/application_layer_spec.cr]: `OkAction`

File: spec/app/application_layer_spec.cr [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+    "OK!"
+  end
+end
+
+MyApp.configure_routes do |routes|
+  routes.add "get", "/ok", OkController
+end
+
+describe "Application Layer" do
+  it "routes to controller" do
```
Comment (sdogruyol) [inline spec/app/application_layer_spec.cr]: `routes to action`

File: src/kemal/app/router.cr [hunk 5]
```diff
@@ -0,0 +1,17 @@
+module Kemal
+  class Router
+    include Kemal::DSL
+
+    def add(method, path, controller_class)
```
Comment (sdogruyol) [inline src/kemal/app/router.cr]: `action_class`

File: src/kemal/app/router.cr [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+      action(method, path) do |context|
+        controller_class.new(context).process
+      end
+    end
+
+    def socket(path, controller_class)
+      ws(path) do |socket, context|
+        controller_class.new(socket, context).process
```
Comment (sdogruyol) [inline src/kemal/app/router.cr]: `action_class`

File: src/kemal/dsl/dsl.cr [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+    HTTP_METHODS   = %w(get post put patch delete options)
+    FILTER_METHODS = %w(get post put patch delete options all)
+
+    {% for method in HTTP_METHODS %}
+      def {{method.id}}(path, &block : HTTP::Server::Context -> _)
+        action({{method}}, path, &block)
+      end
+    {% end %}
+
+    def action(action_name, path, &block : HTTP::Server::Context -> _)
```
Comment (sdogruyol) [inline src/kemal/dsl/dsl.cr]: `action_name` should be `method`

**Comment by sdogruyol** [comment]
Hey @f @askn, WDYT about this?

**Comment by sdogruyol** [comment]
@tarvit there's a failing spec

```
Failures:

  1) Application Layer returns empty response when invalid path
     Failure/Error: client_response.body.should eq("")

       Expected: ""
            got: "    <!DOCTYPE html>\n    <html>\n    <head>\n      <style type=\"text/css\">\n      body { text-align:center;font-family:helvetica,arial;font-size:22px;\n        color:#888;margin:20px}\n      img { max-width: 579px; width: 100%; }\n      #c {margin:0 auto;width:500px;text-align:left}\n      </style>\n    </head>\n    <body>\n      <h2>Kemal doesn't know this way.</h2>\n      <img src=\"/__kemal__/404.png\">\n    </body>\n    </html>"

     # spec/app/application_layer_spec.cr:26

Finished in 16.36 milliseconds
96 examples, 1 failures, 0 errors, 0 pending
```

**Comment by tarvit** [comment]
looks like order dependency issue. will fix that

**Comment by tarvit** [comment]
> @tarvit there's a failing spec

I don't get this problem in my repo. What may be wrong? I run `shards` & `crystal spec`. Do I need to setup something else? Or some custom ENV variables?

BTW I am starting to recieve this 404 response when adding `Kemal.error(404) { |env| render_404 } ` somewhere. But this global setting crashes few other existing tests. 
@sdogruyol could give me some ideas how to resolve this properly?

**Comment by straight-shoota** [comment]
I don't think this kind of indirection between paths and code through a router keeps up with Kemals objective to be "super simple". This makes it unnecessary hard to follow the flow of request handling. That stuff might be useful for heavier applications but should better live in another shard based on Kemal (similar to [sinatra-router](https://github.com/brandur/sinatra-router) for Sinatra).
As suggested in [#373#issuecomment-312883557](#373 (comment)) I'd rather like an approach similar to Sinatra's `Sinatra::Base`/`Sinatra::Application` (see *Modular Apps* in [sinatrarb.com/intro.html](http://www.sinatrarb.com/intro.html)).

So the above example could look like this:

```crystal
# load Kemal without DSL in a global context
require "kemal/base"

class MyApp < Kemal::Application
  get "/ok" do |context|
    puts context.inspect
    "OK!"
  end
end

MyApp.run
```

Way simpler and easily understandable.

**Comment by sdogruyol** [comment]
For this PR, @tarvit how about removing all the special `Action` and `Router` specific stuff and just name spacing the DSL for the starters?

**Comment by tarvit** [comment]
> For this PR, @tarvit how about removing all the special Action and Router specific stuff and just name spacing the DSL for the starters?

Will do. What about naming of a file to require. 
Do you prefer `require "kemal/base"` instead of `require "kemal/main"` ?

**Comment by sdogruyol** [comment]
Thanks @tarvit , I prefer `kemal/base`

**Comment by tarvit** [comment]
I've modified the implementation. So currently DSL looks like this:
```crystal
# load Kemal without DSL in a global context
require "kemal/base"

class MyApp < Kemal::Application
  get "/ok" do |context|
    puts context.inspect
    "OK!"
  end
end

Kemal.run
```
`Kemal::Config::INSTANCE` and `Kemal::RouteHandler::INSTANCE` are still global so it doesn't make sense to add `Kemal::Application#run`

**Comment by sdogruyol** [comment]
Closing this in favor of #378
